Enforce global type declaration for `defaultOptions` that influence results (#13132)
apollographql/apollo-client: .changeset/weak-carrots-give.md

Reviewer: It would be great to mention the signature style in here as well (or perhaps add a separate changeset for it)

---
"@apollo/client": minor
---

Synchronize method and hook return types with `defaultOptions`.

Prior to this change, the following code snippet would always apply:

```ts
declare const MY_QUERY: TypedDocumentNode<TData, TVariables>;
const result1 = useSuspenseQuery(MY_QUERY);
result1.data;
//      ^? TData
const result2 = useSuspenseQuery(MY_QUERY, { errorPolicy: "all" });
result2.data;
//      ^? TData | undefined
```

> While this was generally correct, if you were to register `errorPolicy: 'all'` as a default option for `MY_QUERY`, the type of `result.data` in the first case would still be `TData`, which is not correct - in reality it could also be `undefined`.

Reviewer: ```suggestion
While this was generally correct, if you were to register `errorPolicy: 'all'` as a default option for all queries, the type of `result.data` in the first case would still be `TData`, which is not correct - in reality it could also be `undefined`.
```

Might be best not to point this to this single query as this might confuse some.

> Now, with this change, we are now enforcing that certain `defaultOptions` need to be registered globally on a type level.

Reviewer: ```suggestion
With this change, we are now enforcing that certain `defaultOptions` need to be registered globally on a type level.
```

This means that if you want to use `errorPolicy: 'all'` as a default option for a query, you will need to register it globally like this:

```ts
// apollo.d.ts

> import type {} from "@apollo/client";

Reviewer: ```suggestion
import "@apollo/client";
```

Out of curiosity, do we need the empty import block? I've used a plain import like this with success in the past (something we recommend in our [data masking docs](https://www.apollographql.com/docs/react/data/fragments#setting-data-masking-types))

Author: Good call, changed that everywhere.

---
"@apollo/client": minor
---

Synchronize method and hook return types with `defaultOptions`.

Prior to this change, the following code snippet would always apply:

```ts
declare const MY_QUERY: TypedDocumentNode<TData, TVariables>;
const result1 = useSuspenseQuery(MY_QUERY);
result1.data;
//      ^? TData
const result2 = useSuspenseQuery(MY_QUERY, { errorPolicy: "all" });
result2.data;
//      ^? TData | undefined
```

> While this was generally correct, if you were to register `errorPolicy: 'all'` as a default option for all queries, the type of `result.data` in the first case would still be `TData`, which is not correct - in reality it could also be `undefined`.

Reviewer: ```suggestion
While this was generally correct, if you were to register `errorPolicy: 'all'` as a default option for all queries, the type of `result.data` in the first case would still be `TData`, which is not correct - in reality it could also be `undefined` at runtime.
```

Clarifying that `undefined` means a runtime value here to ensure this point is hammered into the brain of the reader.

        // possible global-registered values:
        // * `errorPolicy`
      }
      interface Mutate {
        // possible global-registered values:
        // * `errorPolicy`
      }
    }
  }
}
```

Once this is in place, the type of `result.data` in the first case will correctly be `TData | undefined`, reflecting the fact that if an error occurs, `data` will be `undefined`.
If you were to manually call `useSuspenseQuery(MY_QUERY, { errorPolicy: 'none' });`, it would become `TData` again.

This change means that you are now forced to register these default options globally, or TypeScript will error out.

Without this global registration, the following (previously valid) code will now error:

```ts
new ApolloClient({
  link: ApolloLink.empty(),
  cache: new InMemoryCache(),
  defaultOptions: {
    watchQuery: {
      // results in a type error:
      // Type '"all"' is not assignable to type '"A default option for watchQuery.errorPolicy must be declared in ApolloClient.DeclareDefaultOptions before usage. See https://www.apollographql.com/docs/react/data/typescript#declaring-default-options-for-type-safety."'.
      errorPolicy: "all",
    },
  },
});
```

> If you are creating multiple instances of Apollo Client with conflicting default options and you cannot register a single `defaultOptions` value as a result, you can relax the constraints by declaring those options as union types covering all values you use. The properties can be required (to enforce them in `defaultOptions`) or optional (if some constructor calls won't pass a value):

Reviewer: Could you also use TypeScript [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) as well to target specific files with these settings? I know in practice that might be more complicated, but might be worth a mention in here as a possible alternative if a targeted approach is needed.

Author: We can mention it, but I hope that's never gonna be the final straw for someone to migrate over to those :D
